File: leetcode/1071-Greatest-Common-Divisor-of-Strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if str1 + str2 != str2 + str1:
            return ""

        # gcd algorithm
        a, b = max(len(str1), len(str2)), min(len(str1), len(str2))
        n = a % b
        while n:
            a = b
            b = n
            n = a % b

        return str1[:b]

File: leetcode/1071-Greatest-Common-Divisor-of-Strings.cpp
class Solution {
public:
    string gcdOfStrings(string str1, string str2) {
        if (str1 + str2 != str2 + str1) return "";

        int a = max(str1.length(), str2.length());
        int b = min(str1.length(), str2.length());
        int n = a % b;

        while (n) {
            a = b;
            b = n;
            n = a % b;
        }
        return str1.substr(0, b);
    }
    
};